File: examples/scripts/topdown.py
#!/bin/python
""" Script for deriving topdown metrics from a Json file produced by Caliper. """

from __future__ import print_function
import sys
import pandas as pd

pd.set_option('display.expand_frame_repr', False)

METRICS = [
    'retiring',
    'bad_speculation',
    'frontend_bound',
    'backend_bound',
    'branch_mispredict',
    'machine_clear',
    'frontend_latency',
    'frontend_bandwidth',
    'memory_bound',
    'core_bound',
    'mem_bound',
    'l1_bound',
    'l2_bound',
    'l3_bound',
    'uncore_bound',
]


def eprint(*args, **kwargs):
    """ Print to stderr """

    print(*args, file=sys.stderr, **kwargs)


def derive_topdown_ivb(dfm):
    """ Perform topdown metric calculations for ivybridge architecture """

    dfm['TEMPORARY_clocks'] = dfm['libpfm.counter.CPU_CLK_UNHALTED.THREAD_P']
    dfm['TEMPORARY_slots'] = 4*dfm['TEMPORARY_clocks']

    # Level 1 - Not Stalled
    dfm['retiring'] = (dfm['libpfm.counter.UOPS_RETIRED.RETIRE_SLOTS']
                       / dfm['TEMPORARY_slots'])
    dfm['bad_speculation'] = ((dfm['libpfm.counter.UOPS_ISSUED.ANY']
                               - dfm['libpfm.counter.UOPS_RETIRED.RETIRE_SLOTS']
                               + 4*dfm['libpfm.counter.INT_MISC.RECOVERY_CYCLES'])
                              / dfm['TEMPORARY_slots'])

    # Level 1 - Stalled
    dfm['frontend_bound'] = (dfm['libpfm.counter.IDQ_UOPS_NOT_DELIVERED.CORE']
                             / dfm['TEMPORARY_slots'])
    dfm['backend_bound'] = (1 - (dfm['frontend_bound']
                                 + dfm['bad_speculation']
                                 + dfm['retiring']))

    # Level 2 - Retiring
    # TODO: implement if possible

    # Level 2 - Bad speculation
    dfm['branch_mispredict'] = (dfm['libpfm.counter.BR_MISP_RETIRED.ALL_BRANCHES']
                                / (dfm['libpfm.counter.BR_MISP_RETIRED.ALL_BRANCHES']
                                   + dfm['libpfm.counter.MACHINE_CLEARS.COUNT']))
    dfm['machine_clear'] = (1 - dfm['branch_mispredict'])  # FIXME: is this correct?

    # Level 2 - Frontend Bound
    dfm['frontend_latency'] = (dfm['libpfm.counter.IDQ_UOPS_NOT_DELIVERED.CORE'].clip(lower=4)
                                     / dfm['TEMPORARY_clocks'])
    dfm['frontend_bandwidth'] = (1 - dfm['frontend_latency'])  # FIXME: is this correct?

    # Level 2 - Backend Bound
    dfm['memory_bound'] = (dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_LDM_PENDING']
                           / dfm['TEMPORARY_clocks'])
    dfm['TEMPORARY_be_bound_at_exe'] = ((dfm['libpfm.counter.CYCLE_ACTIVITY.CYCLES_NO_EXECUTE']
                                         + dfm['libpfm.counter.UOPS_EXECUTED.CORE_CYCLES_GE_1']
                                         - dfm['libpfm.counter.UOPS_EXECUTED.CORE_CYCLES_GE_2'])
                                        / dfm['TEMPORARY_clocks'])
    dfm['core_bound'] = (dfm['TEMPORARY_be_bound_at_exe']
                         - dfm['memory_bound'])

    # Level 3 - Memory bound
    dfm['TEMPORARY_l3_hit_fraction'] = (dfm['libpfm.counter.MEM_LOAD_UOPS_RETIRED.L3_HIT'] /
                                        (dfm['libpfm.counter.MEM_LOAD_UOPS_RETIRED.L3_HIT']
                                         + 7*dfm['libpfm.counter.MEM_LOAD_UOPS_RETIRED.L3_MISS']))
    dfm['TEMPORARY_l3_miss_fraction'] = (7*dfm['libpfm.counter.MEM_LOAD_UOPS_RETIRED.L3_MISS']
                                         / (dfm['libpfm.counter.MEM_LOAD_UOPS_RETIRED.L3_HIT']
                                            + 7*dfm['libpfm.counter.MEM_LOAD_UOPS_RETIRED.L3_MISS']))
    dfm['mem_bound'] = (dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_L2_PENDING']
                        * dfm['TEMPORARY_l3_miss_fraction']
                        / dfm['TEMPORARY_clocks'])
    dfm['l1_bound'] = ((dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_LDM_PENDING']
                        - dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_L1D_PENDING'])
                       / dfm['TEMPORARY_clocks'])
    dfm['l2_bound'] = ((dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_L1D_PENDING']
                        - dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_L2_PENDING'])
                       / dfm['TEMPORARY_clocks'])
    dfm['l3_bound'] = (dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_L2_PENDING']
                       * dfm['TEMPORARY_l3_hit_fraction']
                       / dfm['TEMPORARY_clocks'])
    dfm['uncore_bound'] = (dfm['libpfm.counter.CYCLE_ACTIVITY.STALLS_L2_PENDING']
                           / dfm['TEMPORARY_clocks'])

    for column in dfm.columns:
        if 'TEMPORARY' in column:
            del dfm[column]
        elif 'libpfm.counter' in column:
            del dfm[column]

    return dfm


def derive_topdown(dfm, arch):
    """ Determine topdown function to use, use it, then clean up the dataframe """

    if arch == 'ivybridge':
        dfm = derive_topdown_ivb(dfm)
    else:
        eprint("Error, unsupported architecture " + arch)
        return dfm

    return dfm


def max_column(row, columns):
    """ Returns the column name for the column with the largest value in row """

    return max([(column, row[column]) for column in columns],
               key=lambda t: t[1])[0]


def percentage_string(val):
    """ Returns a percentage-formatted string for a value, e.g. 0.9234 becomes 92.34% """

    return '{:,.2%}'.format(val)


def determine_boundedness(row):
    """ Determine the boundedness of a single row with topdown metrics """

    boundedness = []

    level_1 = max_column(row, ['retiring',
                               'bad_speculation',
                               'frontend_bound',
                               'backend_bound'])
    if str(row[level_1]) != 'nan' and str(row[level_1]) != 'inf':
        boundedness.append(level_1 + ' ' + percentage_string(row[level_1]))

    if level_1 == 'bad_speculation':
        level_2 = max_column(row, ['branch_mispredict',
                                   'machine_clear'])
        boundedness.append(level_2 + ' ' + percentage_string(row[level_2]))
    elif level_1 == 'frontend_bound':
        level_2 = max_column(row, ['frontend_latency',
                                   'frontend_bandwidth'])
        boundedness.append(level_2 + ' ' + percentage_string(row[level_2]))
    elif level_1 == 'backend_bound':
        level_2 = max_column(row, ['core_bound',
                                   'memory_bound'])
        boundedness.append(level_2 + ' ' + percentage_string(row[level_2]))
        if level_2 == 'memory_bound':
            level_3 = max_column(row, ['l1_bound',
                                       'l2_bound',
                                       'l3_bound',
                                       'mem_bound',
                                       'uncore_bound'])
            boundedness.append(level_3 + ' ' + percentage_string(row[level_3]))

    if len(boundedness) == 0:
        boundedness.append('undetermined')

    return boundedness


def analyze_topdown_metrics(dfm):
    """ Analyze topdown metrics to determine boundedness of different Caliper regions """

    dfm['boundedness'] = dfm.apply(determine_boundedness, axis=1)

    for metric in METRICS:
        del dfm[metric]

    return [dict([col for col in row.items() if str(col[1]) != 'nan'])
            for row in dfm.to_dict('index').values()]


def main():
    """ Print all Caliper entries with their derived metrics """

    if len(sys.argv) != 3:
        eprint("Usage: " + sys.argv[0] + " <json file> <arch, e.g. ivybridge>")

    dfm = pd.read_json(sys.argv[1])

    dfm = derive_topdown(dfm, sys.argv[2])

    analysis = analyze_topdown_metrics(dfm)

    for row in analysis:
        print(row)


if __name__ == "__main__":
    main()

File: src/services/libpfm/topdown.cpp
#include "topdown.h"

#include <cmath>
#include <string>

using namespace cali;
using namespace std;

TopdownObject::TopdownObject(std::string &arch) {

    // Initialize event list and derivation metric functions for architecture
    if (arch.compare("sandybridge") == 0) {

    } else if (arch.compare("ivybridge") == 0) {

        event_list = {
            "BR_MISP_RETIRED.ALL_BRANCHES",
            "CPU_CLK_UNHALTED.THREAD_P",
            "CYCLE_ACTIVITY.CYCLES_NO_EXECUTE",
            "CYCLE_ACTIVITY.STALLS_L1D_PENDING",
            "CYCLE_ACTIVITY.STALLS_L2_PENDING",
            "CYCLE_ACTIVITY.STALLS_LDM_PENDING",
            "IDQ.MS_UOPS",
            "IDQ_UOPS_NOT_DELIVERED.CORE",
            "INT_MISC.RECOVERY_CYCLES",
            "MACHINE_CLEARS.COUNT",
            "MEM_LOAD_UOPS_RETIRED.L3_HIT",
            "MEM_LOAD_UOPS_RETIRED.L3_MISS",
            "RESOURCE_STALLS.SB",
            "RS_EVENTS.EMPTY_CYCLES",
            "UOPS_EXECUTED.THREAD",
            "UOPS_EXECUTED.CORE_CYCLES_GE_1",
            "UOPS_EXECUTED.CORE_CYCLES_GE_2",
            "UOPS_ISSUED.ANY",
            "UOPS_RETIRED.RETIRE_SLOTS"
        };

        derivationFunction= {
            [=](SnapshotRecord *snapshot) { 
                std::map<std::string, double> ret;

                double clocks =  
                    getTopdownEventValue(snapshot, "CPU_CLK_UNHALTED.THREAD_P");
                double slots = 
                    4.0 * clocks;
                double retire_slots =  
                    getTopdownEventValue(snapshot, "UOPS_RETIRED.RETIRE_SLOTS");
                ret["retiring"] = 
                    retire_slots / slots;
                ret["bad_speculation"] = 
                    ((getTopdownEventValue(snapshot, "UOPS_ISSUED.ANY")
                            - getTopdownEventValue(snapshot, "UOPS_RETIRED.RETIRE_SLOTS")
                            + 4.0 * getTopdownEventValue(snapshot, "INT_MISC.RECOVERY_CYCLES"))
                        / slots);
                ret["frontend_bound"] = 
                    getTopdownEventValue(snapshot, "IDQ_UOPS_NOT_DELIVERED.CORE") / slots;
                ret["backend_bound"] = 
                    (1.0 - (ret["frontend_bound"]
                            + ret["bad_speculation"]
                            + ret["retiring"]));
                ret["branch_mispredict"] = 
                    getTopdownEventValue(snapshot, "BR_MISP_RETIRED.ALL_BRANCHES") 
                    / (getTopdownEventValue(snapshot, "BR_MISP_RETIRED.ALL_BRANCHES")
                            + getTopdownEventValue(snapshot, "MACHINE_CLEARS.COUNT"));
                ret["machine_clear"] = 1.0 - ret["branch_mispredict"];
                ret["frontend_latency"] =
                    std::max(getTopdownEventValue(snapshot, "IDQ_UOPS_NOT_DELIVERED.CORE"), 4.0)
                        / clocks;
                ret["frontend_bandwidth"] = (1.0 - ret["frontend_latency"]);
                ret["memory_bound"] = 
                    getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_LDM_PENDING")
                    / clocks;
                double be_bound_at_exe = 
                    ((getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.CYCLES_NO_EXECUTE")
                      + getTopdownEventValue(snapshot, "UOPS_EXECUTED.CORE_CYCLES_GE_1")
                      - getTopdownEventValue(snapshot, "UOPS_EXECUTED.CORE_CYCLES_GE_2"))
                     / clocks);
                ret["core_bound"] = 
                    be_bound_at_exe - ret["memory_bound"];
                double l3_hit_fraction = 
                    getTopdownEventValue(snapshot, "MEM_LOAD_UOPS_RETIRED.L3_HIT") /
                    (getTopdownEventValue(snapshot, "MEM_LOAD_UOPS_RETIRED.L3_HIT")
                     + 7*getTopdownEventValue(snapshot, "MEM_LOAD_UOPS_RETIRED.L3_MISS"));
                double l3_miss_fraction = 
                    (7*getTopdownEventValue(snapshot, "MEM_LOAD_UOPS_RETIRED.L3_MISS")
                     / (getTopdownEventValue(snapshot, "MEM_LOAD_UOPS_RETIRED.L3_HIT")
                         + 7*getTopdownEventValue(snapshot, "MEM_LOAD_UOPS_RETIRED.L3_MISS")));
                ret["mem_bound"] = 
                    getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_L2_PENDING")
                        * l3_miss_fraction
                        / clocks;
                ret["l1_bound"] =
                    ((getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_LDM_PENDING")
                      - getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_L1D_PENDING"))
                     / clocks);
                ret["l2_bound"] =
                    ((getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_L1D_PENDING")
                      - getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_L2_PENDING"))
                     / clocks);
                ret["l3_bound"] = 
                    getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_L2_PENDING")
                        * l3_hit_fraction
                        / clocks;
                ret["uncore_bound"] = 
                    getTopdownEventValue(snapshot, "CYCLE_ACTIVITY.STALLS_L2_PENDING")
                        / clocks;

                return ret;
            }
        };
    } else if (arch.compare("broadwell") == 0) {

    } else if (arch.compare("haswell") == 0) {

    } else {

    }

}

void 
TopdownObject::createTopdownEventAttrMap(std::map<std::string, cali::Attribute> attrMap) {
    for (auto const& attrPair : attrMap) {
        for (auto event : event_list) {
            // If event is a substring of this attribute
            std::string attrName = attrPair.first;
            if (attrName.find(event) != std::string::npos) {
                // If event is not assigned, or if event is already assigned but current attribute is longer (e.g. sum) then reassign
                std::map<std::string, Attribute>::iterator assignedEventName = topdownEventsAttrMap.find(event);
                if (assignedEventName == topdownEventsAttrMap.end() || assignedEventName->first.length() < attrName.length()) { // TODO make this look for sum specifically
                    topdownEventsAttrMap[event] = attrPair.second;
                } 
            } 
        }
    }
}

void 
TopdownObject::createTopdownMetricAttrMap(cali::Caliper *c) {
    topdownMetricsAttrMap["retiring"] = c->create_attribute("libpfm.topdown#retiring", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["bad_speculation"] = c->create_attribute("libpfm.topdown#bad_speculation", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["frontend_bound"] = c->create_attribute("libpfm.topdown#frontend_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["backend_bound"] = c->create_attribute("libpfm.topdown#backend_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["branch_mispredict"] = c->create_attribute("libpfm.topdown#branch_mispredict", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["machine_clear"] = c->create_attribute("libpfm.topdown#machine_clear", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["frontend_latency"] = c->create_attribute("libpfm.topdown#frontend_latency", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["frontend_bandwidth"] = c->create_attribute("libpfm.topdown#frontend_bandwidth", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["memory_bound"] = c->create_attribute("libpfm.topdown#memory_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["core_bound"] = c->create_attribute("libpfm.topdown#core_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["mem_bound"] = c->create_attribute("libpfm.topdown#mem_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["l1_bound"] = c->create_attribute("libpfm.topdown#l1_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["l2_bound"] = c->create_attribute("libpfm.topdown#l2_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["l3_bound"] = c->create_attribute("libpfm.topdown#l3_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
    topdownMetricsAttrMap["uncore_bound"] = c->create_attribute("libpfm.topdown#uncore_bound", CALI_TYPE_DOUBLE, CALI_ATTR_ASVALUE);
}

double 
TopdownObject::getTopdownEventValue(cali::SnapshotRecord *snapshot, std::string name) {
    return snapshot->get(topdownEventsAttrMap[name]).value().to_double();
}

void
TopdownObject::addDerivedMetricsToSnapshot(cali::Caliper *c, cali::SnapshotRecord *snapshot) {

    std::map<string, double> derivedMetrics = derivationFunction(snapshot);

    std::vector<cali::Attribute> attr;
    std::vector<cali::Variant>   data;

    for (auto it : derivedMetrics) {
        if (topdownMetricsAttrMap.find(it.first) == topdownMetricsAttrMap.end()) {
            std::cerr << "cant find attribute " << it.first << std::endl;
        } else {
            attr.push_back(topdownMetricsAttrMap[it.first]);

            double val = it.second;
            if (std::isnan(it.second))
                val = 0.0;
            else if (std::isinf(it.second))
                val = 1.0;
            data.push_back(val);
        }
    }

    if (attr.size() > 0)
        c->make_entrylist(attr.size(), attr.data(), data.data(), *snapshot);
}

